scripts/core/helpers/network.tsx

```diff
@@ -0,0 +1,139 @@
+import React from 'react';
+import ng from 'core/services/ng';
+import {gettext} from 'core/utils';
+import {logger} from 'core/services/logger';
+import {appConfig} from '../../appConfig';
+import {showModal} from 'core/services/modalService';
+import {ModalHeader} from 'core/ui/components/Modal/ModalHeader';
+import {ModalBody} from 'core/ui/components/Modal/ModalBody';
+import {ModalFooter} from 'core/ui/components/Modal/ModalFooter';
+import {Modal} from 'core/ui/components/Modal/Modal';
+
+interface IHttpRequestOptions {
+    method: 'GET' | 'POST' | 'PATCH' | 'PUT' | 'DELETE';
+    url: string; // absolute url
+    payload?: {};
+    headers?: {[key: string]: any};
+}
+
+interface IHttpRequestOptionsLocal extends Omit<IHttpRequestOptions, 'url'> {
+    path: string; // relative to application server
+}
+
+interface IHttpRequestJsonOptionsLocal extends IHttpRequestOptionsLocal {
+    // JSON not available with DELETE method
+    method: 'GET' | 'POST' | 'PATCH' | 'PUT';
+}
+
+interface IHttpLocalApiErrorResponse {
+    _error: {
+        code: number;
+        message: string;
+    };
+    _issues: {[key: string]: any};
+    _status: 'ERR' | string;
+}
+
+export function isHttpApiError(x): x is IHttpLocalApiErrorResponse {
+    return typeof x === 'object' && x['_status'] === 'ERR';
+}
+
+function getNetworkErrorModal(message: string) {
+    return class NetworkErrorModal extends React.PureComponent<{closeModal(): void}> {
+        render() {
+            return (
+                <Modal>
+                    <ModalHeader onClose={this.props.closeModal}>
+                        {gettext('Network error occured')}
+                    </ModalHeader>
+                    <ModalBody>
+                        {message}
+                    </ModalBody>
+                    <ModalFooter>
+                        <button
+                            onClick={this.props.closeModal}
+                            className="btn btn-default"
+                            type="button"
+                        >
+                            {gettext('Close')}
+                        </button>
+                    </ModalFooter>
+                </Modal>
+            );
+        }
+    };
+}
+
+function displayErrorMessage(res?: any) {
+    showModal(getNetworkErrorModal(
+        typeof res === 'object' && typeof res._error === 'object' && typeof res._error.message === 'string'
+            ? res._error.message
+            : gettext('An error occured. Retry the action or reload the page.'),
+    ));
+}
+
+function httpRequestBase(options: IHttpRequestOptions): Promise<Response> {
+    const {method, url, payload, headers} = options;
+
+    return fetch(url, {
+        method,
+        headers: headers || {},
+        mode: 'cors',
+        body: typeof payload === 'undefined' ? undefined : JSON.stringify(payload),
```

Comment:
I think the ternary is unnecessary here, what's the difference between this and `body: JSON.stringify(payload)`?

Reply:
You are right, there's none. I thought `JSON.stringify` returns an `'undefined'` string as it does with `null` which turns out isn't true.